blogpost/blogpost.md

```diff
@@ -0,0 +1,845 @@
+        .withUri(env.getProperty("mongo.host")));
+
+    return pipeline.run();
+  }
+
+  private void applyCountStrokeTheGroundLightnings(PCollection<Document> pc) {
+    Integer windowSize = env.getProperty("beam.window.size", Integer.class);
+    pc
+      .apply("Filter hit the ground ones", Filter.by(document -> document.getBoolean("strokeTheGround")))
+      .apply("Add timestamp", WithTimestamps.of(document -> Instant.now()))
+      .apply("Apply fixed windows", Window.into(FixedWindows.of(Duration.standardSeconds(windowSize))))
+      .apply(
+          "Count lightnings that stroke the ground this minute",
+          Combine.globally(Count.<Document>combineFn()).withoutDefaults())
+      .apply("Map to Mongo Document", ParDo.of(new CreateStrikesJson(windowSize)))
+      .apply("Write strikes count to database", MongoDbIO.write()
+          .withDatabase(env.getProperty("mongo.database"))
+          .withCollection(env.getProperty("mongo.collection.strikes"))
+          .withUri(env.getProperty("mongo.host")));
+  }
+
+  private static class CreateStrikesJson extends DoFn<Long, Document> {
+    private final long seconds;
+
+    CreateStrikesJson(long seconds) {
+      this.seconds = seconds;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext ctx) {
+      Document document = new Document()
+          .append("strikes", ctx.element())
+          .append("from", ctx.timestamp().minus(Duration.standardSeconds(seconds)).getMillis())
+          .append("to", ctx.timestamp().getMillis());
+      log.info("Writing strikes {} to database.", document);
+      ctx.output(document);
+    }
+  }
+}
+```
+
+This Bean reads from the Apache Kafka instance to MongoDB.
+
+Before applying the KafkaIO.Read I want to set `.withMaxNumRecords()` conditionally. It allows me to choose whether I want a batch (with maxNumRecords set) or a streaming pipeline. It will be very useful for testing.
+
+Next the json string is parsed to MongoDB bson Document. This is the class required to write to MongoDB.
+The last step is to write the `Document` with lightning data to MongoDB using Beam's MongoDbIO.Write. Just apply it as everything else to the Mongo Collection 'lightnings'.
+
+But why stop there? Let's add an another branch to the pipeline tree starting with `applyCountStrokeTheGroundLightnings(jsonLightningData)`. You can do such a thing to every PCollection object.
+
+This one is a bit more complicated and uses more of Beam's features.
+At first it filters out the lightnings that did not strike the ground (they discharged in the sky) using the `Filter.by(predicate)` transform.
+
+Next it applies a timestamp to each records using `WithTimestamps.of` function.
+Then the streaming input gets windowed to 1 minute windows with `Window.into(FixedWindows.of(duration))`.
+
+The windows are evaluated in combine transform `Combine.globally(Count.<Document>combineFn()).withoutDefaults()`. It counts every element and return the sum.
```

Comment:
Myślę, że warto dodać `for each window`: 

> It counts every element for each window

Reply:
Dzięki!